daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/processors/parsers/PState.scala

```diff
@@ -107,8 +107,11 @@ object MPState {
 class MPState private () {
 
   val arrayIndexStack = MStackOfLong()
+  val arrayCountStack = MStackOfLong()
```

Comment:
I'm hoping we don't need this arrayCountStack (see other comment), but if we do, then I think we need to modify the `Mark` class in the `MPState` object at the top of this file to be able to capture and reset this state in case backtracking occurs.

Reply:
Scaladoc this element to explain exactly why it is needed, what it is counting, etc.

Reply:
I do see a need for this counter, as the array index is needed to keep track of where in the iteration space we are, for maxOccurs handling, defaulting, etc. 
The count handles the fact that despite all that iteration, one might not actually be attaching anything to the infoset.

So I do think you will need to add the arrayCountStack to the save/restore state stuff for backtracking.

Reply:
>  maxOccurs 

Isn't maxOccurs just about what appears in the infoset?

> defaulting

If something is defaulted, doesn't that mean it will have an instance in the infoset and so would be counted in arrayIndex?

Reply:
maxOccurs bounds what appears in the infoset. 

We need to count the number of iterations of the parsing. Why? because of maxOccurs and the forward progress requirement. If maxOccurs is bounded, then even if we aren't adding anything to the infoset we still iterate maxOccurs times. So we need the counter that has nothing to do with how many infoset items have actually been retained and attached to the infoset. Just so we know when to stop iterating. 

It's only when maxOccurs is unbounded that we insist on forward progress for each parse.  That forward progress can; however, produce the "empty" representation (e.g., for an empty string element that might be something that requires quotation marks. That would still result in nothing in the infoset if treatAsAbsent is the emptyElementParsePolicy setting. So we can still be consuming data (quotation marks) and yet not adding anything to the infoset. So we can make forward progress in the data without creating any infoset elements. 

So it's pretty clear there are two pieces of state here. One is the index/count of how many infoset elements there are. The other is the iteration count going up to maxOccurs only in the case where that bound is being used. 

(btw: this whole thing probably not a good idea in the DFDL spec, and we probably need to change this despite the DFDL specification. Tools like XSAT insist on large values for maxOccurs rather than unbounded, but that changes the behavior of the schema from one that insists on forward progress to one that does not. I think this was not understood when DFDL was specified, that changing from unbounded to a bound would actually change parse behavior on the very first parse of an array element. So maxOccurs=4000000 means even if the parse is stuck in a rut we'll iterate 4000000 times, where if maxOccurs was unbounded we'd stop immediately. )

So I think adding a second counter, which answers the question how-many infoset elements, distinct from the counter that is providing a count up to maxOccurs to stop the iteration, makes sense to me. 

Interestingly, dfdl:occursIndex function, if called, needs to give back, not the iteration counter, but the index of the current "real" infoset element being attached.

Reply:
Ok, just to be to clear, we could have something like `dfdl:occursCountKind="implicit"`, `minOccurs="5"`, `maxOccurs="5"`, and with the right combination of other settings, we know daffodil will parse 5 things, but some of those could be "empty", and so we could end up with less than 5 things in the infoset? Very confusing, but good to know.

Followup question, do we even need a second counter? In the cases where we need to know the actual number of occurrences in an infoset, can we just count the number of occurrences in the infoset? The `DIArray` class has a `length` member with this information. Can we just use that?

We also probably need to take a good look at `arrayIndexStack` and when it's used. I'm concerned that we make assumptions in a number of places that this is the number of array elements, and that's not always the case. This is for sure the case with `dfdl:occursIndex()` function, which uses arrayPos which I think is just the top of the stack. Maybe `arrayIndexStack` really wants to be renamed to something like `arrayIterationStack` or something, and have some thorough comments about what it is and why it's not necessarily the same as the number of array elements in the infoset.  Fine with me if this is a new issue/PR.

Reply:
Since DIArray already has a member that contains the count, yes I would think we should just use that. 

Changes to the array will already get backtracked properly. 

If dfdl:occursIndex() is returning the iteration number, then that's going to be a bug for sure in these sparse array cases where attempts to parse an element succeed but create an empty value that causes the element to not be attached to the infoset.  Separate bug though. We need a test to illustrate it.

Reply:
I wrote up a quick test and can confirm that `dfdlx:occursIndex()` returns the iteration number and not the number of elements in the infoset. I'll open a ticket and submit a PR to add the test shortly.

Reply:
> In the cases where we need to know the actual number of occurrences in an infoset, can we just count the number of occurrences in the infoset? The `DIArray` class has a `length` member with this information. Can we just use that?

I tried to find an easy way from within the `endArray` function to get a hold of that but was unable to find one. The current node at this point is the containing sequence itself (padRoot, not nz). What is the best way to get a reference to the `DIArray` here?

Reply:
I think this code does something similar to what you need:

https://github.com/apache/daffodil/blob/main/daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/processors/parsers/SequenceParserBases.scala#L216-L238

I think `pstate.infoset.maybeLastChild` should return the last `DINode` child of padRoot (i.e. the nz array). And since the `DINode` interface has a `numChildren` method, you might be able to just use that and not even need to check if it's an array. So if the child isn't even an array validation will still work. I think...

Reply:
Note: DAFFODIL-2791 and PR #964 added to show that `dfdl:ocurrsIndex` (and likely anything else using arrayIndex as a count of infoset items) is probably broken if when `emptyElemenParsePolicy="treatAsMissing"`.

Reply:
> I think pstate.infoset.maybeLastChild should return the last DINode child of padRoot (i.e. the nz array). And since the DINode interface has a numChildren method, you might be able to just use that and not even need to check if it's an array. So if the child isn't even an array validation will still work. I think...

Right above the `endArray` call in the above snippet, `maybeLastChild` at that point is `Nope`, and the contents of `pstate.infoset` contains padRoot with no children. I was trying to obtain this information from the `ElementRuntimeData` but that could contain anything inside the sequence--multiple arrays, simple types, null values, etc.  

And is it me, or does the debugger have issues finding variables or are they getting optimized out? At times I have to do some aliasing to get things to appear, such as `val foo = this.foo`

Reply:
I think that might be correct? All the children of padRoot have been removed in this particular test because they were all empty.  So if `infoset.maybeLastChild` is `Nope`, then the number of occurrences of the array we were parsing is zero. If `infoset.maybeLastChild` is defined, then it should be a `DIArray` instance and numChildren should get the length of that array.

But I think maybe that raises a potential edge case--if there were previous siblings to the array, but the array itself had no occurrences, then `maybeLastChild` would return some DINode other than the array we are trying to validate. So if `maybeLastChild` is defined, then we need to make sure it's the thing we actually care about. Maybe the logic looks something like this:

```scala
val maybeLasChild = infoset.maybeLastChild
val arrayOccurrences =
  if (maybeLastChild.isEmpty || maybeLastChild.get.erd != erd)
    // no occurrences for the array this parser cares about were added
    0
  else 
    maybeLastChild.get.numChildren
```

I don't use the Scala debugger, so unfortunately I can' help with that.

Reply:
Updated to count the children of maybeLastChild